### PR TITLE
Laser Targeting Optimizations

### PR DESCRIPTION
**Pull request by JustWannaFly**
* Lasers now scan for ILaserTargets only when they are placed or when a world update event occurs within a cube extending 6 blocks out from the laser
* When lasers would previously scan for targets, they now just look through their lists of targets to decide which, if any, targets need power and which one to send power to
* There should be no difference in behavior, just saving in processing time

### CODE REVIEW

**Comment by asiekierka** [comment]
You're registering every single tile entity as an IWorldEventListener. By the amount of update calls which can be made in a single tick, I am doubtful as to the performance benefits.

You'd be better off making a single IWorldEventListener, keeping a Map<ChunkPos, TileEntity> and using that. And I'd still want benchmarks in that case.

**Comment by AEnterprise** [comment]
you might also instead of scanning all blocks again only just process that one block that got updated and add/remove only that block (as the rest in range will not have changed)

**Comment by asiekierka** [comment]
@AEnterprise Doesn't VolumeUtil perform non-air block occlusion, though?

**Comment by AEnterprise** [comment]
it does, but the laser could do that as well? because the volumeutil not only needs to iterate over all blocks in the volume but also do the LOS check

maybe a better alternative to make sure you don't check constantly would be when a block in range changes you set a boolean flag so it checks the next tick (so it can't run multiple times per tick)

**Comment by JustWannaFly** [comment]
I hadn't considered that the callback can be triggered multiple times per tick. I personally like the idea of setting a flag and just having the block scan based on the flag. @asiekierka are you alright with that approach to make sure we only scan once?

**Comment by AEnterprise** [comment]
if you do that and the global listener it should give a good boost to laser performance compared to the old code (old ingame profiling was putting 20 lasers or so at 60-70% of laser ticktime)

**Comment by JustWannaFly** [comment]
It very well could be. I'll have to look at that. I'm new to forge so I'm not surprised that there's probably a better way than I came up with.

**Comment by JustWannaFly** [comment]
@AEnterprise I thought of that situation and believe it would be handled by the fact that the WorldUpdateCollector returns a list of updates that took place in the previous tick; however I'll admit that seeing the implementation of a single event listener for all the lasers I'm not convinced it offers much in terms of performance improvements. I'll be pushing a commit shortly that should offer the performance this system would without having the lasers rely on some external event listener.

**Comment by JustWannaFly** [comment]
Looking at how easily the laser can be switched to only scanning for targets once per tick at most, I think that is definitely the way to go. If there is agreement all around I'll remove the WorldUpdateDetector and I think the PR would be ready for merging after that.

**Comment by AEnterprise** [comment]
@JustWannaFly that is exactly what it should do, the main question remaining was if we all lasers should be registered as event listers themselfs or if we should have a global one where the lasers register themselfs in that then sets the flag on the laser if there is a block update in the area

**Comment by JustWannaFly** [comment]
The BlockUpdateDetector was my attempt at creating the global event listener to do what you describe, but after seeing it start to come into existence, it seems far simpler (to me at least) to just let each laser be an event listener.

**Comment by AEnterprise** [comment]
well asie brought up the point of it maybe being more performant, can you maybe try both aproach and benchmark them to see if there is indeed a performance difference?

**Comment by JustWannaFly** [comment]
Sure. Is there a preferred way to see how long a section of Buildcraft takes to run? I haven't attempted to benchmark before and could use a starting point

**Comment by AEnterprise** [comment]
you can try using visualVM or the minecraft debug timing, use the /debug start  command, let it run for a while (preferably with quite a few lasers in different places), stop and open the file to see the stats.

can you do scenario's for before the change, with all lasers registering as listeners and with a global one that forwards to the affected lasers? (mabye in all cases once with not much updates and one with a lot of block updates)?

**Comment by JustWannaFly** [comment]
The results are in! 
**TLDR:** I can't find a discernible difference in performance either way and am ready to finish up the code in whichever style the community finds better.

**Full Report**
I ended adding a little bit of code to the laser to calculate the duration of each tick in nanoseconds and took the average duration over 10000 total laser ticks across lasers for my bench-marking. The code to do that can be seen [in this commit](https://github.com/JustWannaFly/BuildCraft/commit/9b6966038d9dd76ac41a3431e44a030e0af13028)

World Setup (and zip):
384 lasers setup in 2 6*32 arrays on either side of an row of 30 pistons on a toggleable observer clock
[Laser Benchmarking World.zip](https://github.com/BuildCraft/BuildCraft/files/1324638/Laser.Benchmarking.World.zip)

Minimal updates were with the pistons locked, Many updates were with the pistons firing

**Observed FPS:**
(this was just "measured" by turning off V Sync, uncapping the framerate, and watching the number with the F3 screen on. It's not very precise but just to get a general feel)

All lasers listen:
Minimal updates FPS: roughly 400-500 fps
Lots of updates FPS: roughly 220-320 fps

Global Listener:
Minimal updates FPS: roughly 400-500 fps
Lots of updates FPS: roughly 216-321 fps

Average Tick times:
These measurements started with the pistons deactivated, then turned on and then turned off again. The transitions are pretty clear

Here are the averages for each laser being a world listener:
[07:38:36] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 35993.9279 nanoseconds
[07:38:37] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 36058.4171 nanoseconds
[07:38:38] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 35680.9469 nanoseconds
[07:38:40] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 32481.1503 nanoseconds
[07:38:41] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 36153.7704 nanoseconds
[07:38:43] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 121249.8682 nanoseconds
[07:38:45] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 192024.014 nanoseconds
[07:38:47] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 183923.0327 nanoseconds
[07:38:47] [Server thread/WARN]: Can't keep up! Did the system time change, or is the server overloaded? Running 2176ms behind, skipping 43 tick(s)
[07:38:49] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 181556.2737 nanoseconds
[07:38:51] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 172450.6463 nanoseconds
[07:38:52] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 165548.9382 nanoseconds
[07:38:54] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 167486.8377 nanoseconds
[07:38:56] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 171262.2737 nanoseconds
[07:38:56] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 39882.6225 nanoseconds
[07:38:57] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 31700.542 nanoseconds
[07:38:57] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 30473.1434 nanoseconds
[07:38:57] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 30713.4905 nanoseconds
[07:38:58] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 35315.2335 nanoseconds
[07:38:59] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 44450.2138 nanoseconds

Average tick times for the global listener all lasers are registered to:
[08:00:39] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 33588.7095 nanoseconds
[08:00:40] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 33661.9369 nanoseconds
[08:00:41] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 33559.4239 nanoseconds
[08:00:43] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 34409.7544 nanoseconds
[08:00:44] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 33887.1949 nanoseconds
[08:00:45] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 34070.1293 nanoseconds
[08:00:47] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 33496.8378 nanoseconds
[08:00:48] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 32771.9333 nanoseconds
[08:00:50] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 115522.4596 nanoseconds
[08:00:51] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 178492.9665 nanoseconds
[08:00:53] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 175026.9583 nanoseconds
[08:00:55] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 181915.4966 nanoseconds
[08:00:56] [Server thread/WARN]: Can't keep up! Did the system time change, or is the server overloaded? Running 2408ms behind, skipping 48 tick(s)
[08:00:57] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 179588.0299 nanoseconds
[08:00:59] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 173715.4173 nanoseconds
[08:01:01] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 181128.325 nanoseconds
[08:01:03] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 178844.1143 nanoseconds
[08:01:04] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 156400.9598 nanoseconds
[08:01:05] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 32949.2433 nanoseconds
[08:01:05] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 33042.2497 nanoseconds
[08:01:05] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 32947.6541 nanoseconds
[08:01:06] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 33048.0133 nanoseconds
[08:01:07] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 33050.8409 nanoseconds
[08:01:08] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 32766.9655 nanoseconds
[08:01:09] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 33261.9113 nanoseconds
[08:01:10] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 33088.7617 nanoseconds
[08:01:12] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 32730.4186 nanoseconds
[08:01:13] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 33062.0484 nanoseconds
[08:01:14] [Server thread/INFO]: [STDOUT]: Average Laser tick time = 33387.5112 nanoseconds

**Comment by JustWannaFly** [comment]
And now the results are in! (actually measuring the time processing events this time)

TLDR
===

With very few world updates having each laser as a listener seems to be slightly better
With lots of world updates having one global listener is better

Over all, I think having one global listener is more advantageous for performance by a small, but consistent margin in more intense setups

Full Test
====

[Code used to benchmark](https://github.com/JustWannaFly/BuildCraft/commit/856aeb873d889b12566b76aeb5e8cab77c02fee4)

[Test World](https://github.com/BuildCraft/BuildCraft/files/1326640/Laser.Benchmarking.World.zip)

**Setup**
384 lasers setup in 2 6*32 arrays on either side of an row of 30 pistons on a toggleable observer clock

Minimal updates were with the pistons locked, Many updates were with the pistons in motion

All total scanning time was totaled over a period of 10,000 ticks total across all lasers and then reset for the next 10,000 laser ticks

*Note: I realized the console output message is pretty bad as I was analyzing the logs. I didn't think it was worth redoing the testing just for the sake of improving the log message. It is actually processing the time spent processing world updates as can be seen in the code.

**Each Laser Listens for updates**

Minimal updates FPS: roughly 400-500 fps
Lots of updates FPS: roughly 220-320 fps

Total time processing Updates (minimal updates) Range: 0.198ms - 0.612ms

[10:30:58] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 371271 nanoseconds
[10:30:59] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 346706 nanoseconds
[10:31:00] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 323008 nanoseconds
[10:31:02] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 308368 nanoseconds
[10:31:03] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 245184 nanoseconds
[10:31:04] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 260991 nanoseconds
[10:31:05] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 563207 nanoseconds
[10:31:07] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 580734 nanoseconds
[10:31:08] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 541542 nanoseconds
[10:31:09] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 612068 nanoseconds
[10:31:11] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 566998 nanoseconds
[10:31:12] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 507046 nanoseconds
[10:31:13] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 420708 nanoseconds
[10:31:15] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 294640 nanoseconds
[10:31:16] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 292579 nanoseconds
[10:31:17] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 197501 nanoseconds
[10:31:18] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 228510 nanoseconds
[10:31:20] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 217387 nanoseconds
[10:31:21] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 212708 nanoseconds
[10:31:22] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 219729 nanoseconds
[10:31:24] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 197776 nanoseconds


Total time processing Updates (many updates) Range: 11.723ms - 14.852ms

[10:32:21] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 12251105 nanoseconds
[10:32:23] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 14851893 nanoseconds
[10:32:25] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 12469054 nanoseconds
[10:32:26] [Server thread/WARN]: Can't keep up! Did the system time change, or is the server overloaded? Running 2048ms behind, skipping 40 tick(s)
[10:32:26] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 12891347 nanoseconds
[10:32:28] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 12788336 nanoseconds
[10:32:30] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 12258432 nanoseconds
[10:32:31] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 13371746 nanoseconds
[10:32:33] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 12635965 nanoseconds
[10:32:35] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 12254862 nanoseconds
[10:32:36] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 13017561 nanoseconds
[10:32:38] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 12529912 nanoseconds
[10:32:40] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 12038457 nanoseconds
[10:32:41] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 12811514 nanoseconds
[10:32:43] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 12754730 nanoseconds
[10:32:45] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 11723148 nanoseconds
[10:32:46] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 12850108 nanoseconds
[10:32:48] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 12462798 nanoseconds
[10:32:49] [Server thread/WARN]: Can't keep up! Did the system time change, or is the server overloaded? Running 6964ms behind, skipping 139 tick(s)
[10:32:50] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 11738003 nanoseconds
[10:32:51] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 12679861 nanoseconds
[10:32:53] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 12570767 nanoseconds

**One global listener**

Minimal updates rough FPS: 400-500 fps
Lots of updates rough FPS: 216-321 fps

Total time processing Updates (minimal updates) Range: 0.415ms - 2.654ms (the long times were only present when first launching the game. The range after launch is more like 0.415ms - 0.6ms after watching the logs for a while)

[10:26:09] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 1028389 nanoseconds
[10:26:11] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 2111198 nanoseconds
[10:26:12] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 2654211 nanoseconds
[10:26:13] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 1924832 nanoseconds
[10:26:14] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 1739630 nanoseconds
[10:26:16] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 1694871 nanoseconds
[10:26:17] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 1456716 nanoseconds
[10:26:18] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 932429 nanoseconds
[10:26:20] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 912237 nanoseconds
[10:26:21] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 932134 nanoseconds
[10:26:22] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 1074328 nanoseconds
[10:26:23] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 989772 nanoseconds
[10:26:25] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 1021662 nanoseconds
[10:26:26] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 916920 nanoseconds
[10:26:27] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 460802 nanoseconds
[10:26:29] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 420720 nanoseconds
[10:26:30] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 452316 nanoseconds
[10:26:31] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 415744 nanoseconds
[10:26:33] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 423353 nanoseconds
[10:26:34] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 475428 nanoseconds
[10:26:34] [Server thread/INFO]: Saving and pausing game...
[10:26:34] [Server thread/INFO]: Saving chunks for level 'Laser Benchmarking'/Overworld
[10:26:34] [Server thread/INFO]: Saving chunks for level 'Laser Benchmarking'/Nether
[10:26:34] [Server thread/INFO]: Saving chunks for level 'Laser Benchmarking'/The End
[10:27:03] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 443244 nanoseconds
[10:27:04] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 437394 nanoseconds
[10:27:05] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 545349 nanoseconds
[10:27:07] [Server thread/INFO]: Player17 has just earned the achievement [Taking Inventory]
[10:27:07] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 492399 nanoseconds
[10:27:07] [Client thread/INFO]: [CHAT] Player17 has just earned the achievement [Taking Inventory]
[10:27:08] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 544475 nanoseconds
[10:27:09] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 454072 nanoseconds
[10:27:10] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 584554 nanoseconds
[10:27:12] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 433591 nanoseconds
[10:27:13] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 420135 nanoseconds
[10:27:14] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 510541 nanoseconds
[10:27:16] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 436809 nanoseconds
[10:27:17] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 451730 nanoseconds
[10:27:18] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 475722 nanoseconds
[10:27:20] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 461388 nanoseconds
[10:27:21] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 456703 nanoseconds
[10:27:22] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 440906 nanoseconds
[10:27:23] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 471625 nanoseconds
[10:27:25] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 461970 nanoseconds
[10:27:26] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 480109 nanoseconds

Total time processing Updates (many updates) Range: 9.769ms - 11.100ms

[10:27:58] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 10261673 nanoseconds
[10:28:00] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 10963546 nanoseconds
[10:28:02] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 10357938 nanoseconds
[10:28:04] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 11099554 nanoseconds
[10:28:05] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 10826885 nanoseconds
[10:28:05] [Server thread/WARN]: Can't keep up! Did the system time change, or is the server overloaded? Running 2447ms behind, skipping 48 tick(s)
[10:28:07] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 10305277 nanoseconds
[10:28:09] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 10671013 nanoseconds
[10:28:10] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 10836245 nanoseconds
[10:28:12] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 9811985 nanoseconds
[10:28:14] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 10558968 nanoseconds
[10:28:15] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 10330727 nanoseconds
[10:28:17] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 9823095 nanoseconds
[10:28:18] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 10361990 nanoseconds
[10:28:20] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 10142295 nanoseconds
[10:28:22] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 9601631 nanoseconds
[10:28:23] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 10353803 nanoseconds
[10:28:25] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 10286266 nanoseconds
[10:28:27] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 9768972 nanoseconds
[10:28:27] [Server thread/WARN]: Can't keep up! Did the system time change, or is the server overloaded? Running 6511ms behind, skipping 130 tick(s)
[10:28:28] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 10540209 nanoseconds
[10:28:30] [Server thread/INFO]: [STDOUT]: Total Laser scan time = 10184150 nanoseconds

**Comment by JustWannaFly** [comment]
One quick question: would it be worth making this event listener generic so that anything that implements an interface (say IBlockUpdateSubscriber) can receive updates or should I just keep it tied to the laser?

File: common/buildcraft/lib/block/LocalBlockUpdateNotifier.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+     * @param flags from the Block Update
+     */
+    private void notifySubscribersInRange(@Nonnull World world, @Nonnull BlockPos eventPos, @Nonnull IBlockState oldState,
+                                          @Nonnull IBlockState newState, int flags) {
+        for (BlockPos keyPos : subscriberMap.keySet()) {
+            int updateRange = subscriberMap.get(keyPos).getUpdateRange();
+            if (Math.abs(keyPos.getX() - eventPos.getX()) <= updateRange &&
+                    Math.abs(keyPos.getY() - eventPos.getY()) <= updateRange &&
+                    Math.abs(keyPos.getZ() - eventPos.getZ()) <= updateRange) {
+                subscriberMap.get(keyPos).setWorldUpdated(world, eventPos, oldState, newState, flags);
```
Comment (AlexIIL) [inline common/buildcraft/lib/block/LocalBlockUpdateNotifier.java]: I'd recommend using 
````java
for (ILocalBlockUpdateSubscriber subscriber : subscriberSet) {
  BlockPos keyPos = subscriber.getSubscriberPos();
  // Rest of code
}
````
(And change from HashMap<BlockPos, Subscriber> to HashSet<Subscriber> as you can just get the subscriber position from the subscriber itself, and there's no need to map it from a block position)

File: common/buildcraft/lib/block/ILocalBlockUpdateSubscriber.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+     * This method should be kept lightweight as it can be called multiple times per tick.
+     *
+     * @param world from the block update event
+     * @param eventPos from the block update event
+     * @param oldState from the block update event
+     * @param newState from the block update event
+     * @param flags from the block update event
+     */
+    void setWorldUpdated(@Nonnull World world, @Nonnull BlockPos eventPos, @Nonnull IBlockState oldState,
+                         @Nonnull IBlockState newState, int flags);
```
Comment (AlexIIL) [inline common/buildcraft/lib/block/ILocalBlockUpdateSubscriber.java]: Don't bother with all these Nonnull annotations -- I still need to figure out a sensible way of getting default nullness annotations to work with eclipse + idea, and these just add more space.

File: common/buildcraft/lib/block/LocalBlockUpdateNotifier.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+
+/**
+ * Listens for BlockUpdates in a given world and notifies all registered IBlockUpdateSubscribers of the update provided
+ * it was within the update range of the ILocalBlockUpdateSubscriber
+ */
+public class LocalBlockUpdateNotifier {
+
+    private static Map<World, LocalBlockUpdateNotifier> instanceMap = new WeakHashMap<>();
+    private Map<BlockPos, ILocalBlockUpdateSubscriber> subscriberMap = new HashMap<>();
```
Comment (AlexIIL) [inline common/buildcraft/lib/block/LocalBlockUpdateNotifier.java]: Bit minor, but these can probably be declared final.

File: common/buildcraft/silicon/tile/TileLaser.java [hunk 4]
```diff
@@ -47,12 +52,14 @@
 import buildcraft.silicon.BCSiliconBlocks;
 import buildcraft.silicon.client.render.AdvDebuggerLaser;
 
-public class TileLaser extends TileBC_Neptune implements ITickable, IDebuggable {
+public class TileLaser extends TileBC_Neptune implements ITickable, IDebuggable, ILocalBlockUpdateSubscriber {
+    private List<BlockPos> targetPositions;
```
Comment (AlexIIL) [inline common/buildcraft/silicon/tile/TileLaser.java]: Change to final, and use ````clear()```` rather than reallocating a new ````ArrayList```` every time.

File: common/buildcraft/silicon/tile/TileLaser.java [hunk 5]
```diff
@@ -47,12 +52,14 @@
...
     private BlockPos targetPos;
     private final AverageLong avgPower = new AverageLong(100);
     private long averageClient;
     private final MjBattery battery;
-
+    private final int TARGETING_RANGE = 6;
```
Comment (AlexIIL) [inline common/buildcraft/silicon/tile/TileLaser.java]: Change to static + move to the top of the file.

File: common/buildcraft/silicon/tile/TileLaser.java [hunk 6]
```diff
@@ -64,50 +71,75 @@ public TileLaser() {
...
             }
-        } else {
-            return null;
         }
+        return false;
+    }
+
+    private ILaserTarget getTarget() {
+        if (targetPos != null) {
+            return (ILaserTarget) world.getTileEntity(targetPos);
```
Comment (AlexIIL) [inline common/buildcraft/silicon/tile/TileLaser.java]: Don't blindly cast -- check if the returned tile entity actually is an ILaserTarget (it might change before you get updated)

File: common/buildcraft/silicon/tile/TileLaser.java [hunk 7]
```diff
@@ -64,50 +71,75 @@ public TileLaser() {
...
         }
-
-        targetPos = possible.get(world.rand.nextInt(possible.size()));
+        targetPos = targetsNeedingPower.get(world.rand.nextInt(targetsNeedingPower.size()));
     }
 
-    private ILaserTarget getTarget() {
-        if (targetPos != null) {
-            TileEntity tile = world.getTileEntity(targetPos);
+    private boolean powerNeededAt(BlockPos position) {
```
Comment (AlexIIL) [inline common/buildcraft/silicon/tile/TileLaser.java]: Perhaps ````isPowerNeededAt````?

File: common/buildcraft/lib/block/LocalBlockUpdateNotifier.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        subscriberMap.put(subscriber.getSubscriberPos(), subscriber);
+    }
+
+    /**
+     * Stop an @{ILocalBlockUpdateSubscriber} from receiving notifications about block updates
+     * @param subscriber the subscriber to no longer receive notifications about local block update
+     */
+    public void removeSubscriberFromUpdateNotifications(ILocalBlockUpdateSubscriber subscriber) {
+        if (subscriberMap.containsKey(subscriber.getSubscriberPos())) {
+            subscriberMap.remove(subscriber.getSubscriberPos());
```
[the remark this answers is not in the document]
Comment (JustWannaFly) [reply common/buildcraft/lib/block/LocalBlockUpdateNotifier.java]: Forgot that remove does that. It's gone

File: common/buildcraft/lib/block/LocalBlockUpdateNotifier.java (hunk 1 above)
Comment (JustWannaFly) [reply common/buildcraft/lib/block/LocalBlockUpdateNotifier.java]: Agreed on using a set. I just wasn't thinking...

File: common/buildcraft/silicon/tile/TileLaser.java (hunk 7 above)
Comment (JustWannaFly) [reply common/buildcraft/silicon/tile/TileLaser.java]: I like the is but I'm not sure about keeping at. Do you prefer `isPowerNeeded` or `isPowerNeededAt`?

File: common/buildcraft/silicon/tile/TileLaser.java (hunk 4 above)
Comment (JustWannaFly) [reply common/buildcraft/silicon/tile/TileLaser.java]: Oops, I wasn't thinking about the cost of memory allocation while working on a task to improve performance :man_facepalming:

File: common/buildcraft/silicon/tile/TileLaser.java (hunk 7 above)
Comment (AlexIIL) [reply common/buildcraft/silicon/tile/TileLaser.java]: I hadn't actually noticed the "at" first time, and just copy-pasted the existing words without thinking. Woops. Seeing as the method takes a blockposition rather than nothing, having the "at" probably makes sense.

**Comment by AEnterprise** [comment]
if i'm not mistaken this means that we also no longer need the interface on blocks?

**Comment by AlexIIL** [comment]
@AEnterprise we never needed it, but it's still a performance enhancement to have it.

**Comment by AEnterprise** [comment]
only a very minor one, it makes things slightly more complex then without to implement and i don't think we would see that much difference overall since it's updaing less frequently

**Comment by AlexIIL** [comment]
Looks good :)

Syncing laser power to the client seems broken atm, but it looks like that is broken anyway, so I'll fix it separately.
